File: src/mapper.zig
const std = @import("std");
const expect = @import("std").testing.expect;

const MAX_KEY_SIZE = 2048;
const MAX_VALUE_SIZE = 2048;
const HASH_MAP_SIZE = 101;
const NUMBER_OF_MAPS = 100;

const Value = []u8;
const Key = []u8;

pub const Entry = struct {
    key: Key,
    value: Value,
    next: ?*Entry,
};

var maps: [NUMBER_OF_MAPS]?[]?*Entry = undefined;
var maps_count: usize = 0;
// var alloca = std.heap.loggingAllocator(std.heap.c_allocator);
var allocator = std.heap.c_allocator;

fn hash(key: []const u8) usize {
    var hashVal: usize = 0;
    for (key) |c| {
        hashVal += @intCast(c);
    }
    return @mod(hashVal, HASH_MAP_SIZE);
}

export fn init_map() i32 {
    if (maps_count >= NUMBER_OF_MAPS) {
        std.debug.print("[ZIG_MAPPER_ERROR]Maximum number of maps reached\n", .{});
        return -1;
    }
    const map_id: usize = @intCast(maps_count);
    maps_count += 1;
    const allocating = allocator.alloc(?*Entry, HASH_MAP_SIZE);
    if (allocating) |array| {
        for (array) |*entry| {
            entry.* = null;
        }
        maps[map_id] = array;
    } else |err| {
        std.debug.print("[ZIG_MAPPER_ERROR]Failed to allocate memory: {}\n", .{err});
        return -1;
    }

    std.debug.print("[ZIG_MAPPER_INFO]Creating new map {}\n", .{map_id});
    return @intCast(map_id);
}

export fn set_val(map_id: i32, key: [*c]const u8, key_len: usize, value: [*c]const u8, value_len: usize) i32 {
    if (map_id < 0 or map_id >= maps_count or key_len > MAX_KEY_SIZE or value_len > MAX_VALUE_SIZE) {
        std.debug.print("[ZIG_MAPPER_ERROR]Invalid parameters: map_id={0}, key_len={1}\n", .{ map_id, key_len });
        return -1;
    }

    const map = maps[@intCast(map_id)] orelse return -1;
    const hash_value = hash(key[0..key_len]);

    // Try finding an existing entry with the same key.
    var current_entry_ptr = &map[hash_value];
    while (current_entry_ptr.* != null) {
        const current_entry = current_entry_ptr.*;
        if (current_entry.?.key.len == key_len and std.mem.eql(u8, current_entry.?.key[0..key_len], key[0..key_len])) {
            // Allocate new value buffer and handle allocation failure.
            const newValue = allocator.alloc(u8, value_len) catch {
                std.debug.print("[ZIG_MAPPER_ERROR]Failed to allocate memory for value\n", .{});
                return -1;
            };
            std.mem.copy(u8, newValue[0..value_len], value[0..value_len]);
            allocator.free(current_entry.?.value);
            current_entry.?.value = newValue[0..value_len];
            return 0;
        }
        current_entry_ptr = &current_entry.?.next;
    }

    // Allocate and initialize a new entry if no existing entry matches.
    const new_entry = allocator.create(Entry) catch {
        std.debug.print("[ZIG_MAPPER_ERROR]Failed to allocate memory for new entry\n", .{});
        return -1;
    };
    new_entry.* = Entry{
        .key = allocator.alloc(u8, key_len) catch {
            std.debug.print("[ZIG_MAPPER_ERROR]Failed to allocate memory for key\n", .{});
            allocator.destroy(new_entry);
            return -1;
        },
        .value = allocator.alloc(u8, value_len) catch {
            std.debug.print("[ZIG_MAPPER_ERROR]Failed to allocate memory for value\n", .{});
            allocator.free(new_entry.key);
            allocator.destroy(new_entry);
            return -1;
        },
        .next = null,
    };
    std.mem.copy(u8, new_entry.key[0..key_len], key[0..key_len]);
    std.mem.copy(u8, new_entry.value[0..value_len], value[0..value_len]);

    // Insert the new entry into the map.
    new_entry.next = map[hash_value];
    map[hash_value] = new_entry;

    return 0;
}

export fn get_val(map_id: i32, key: [*c]const u8, key_len: i32, value: [*c]u8) i32 {
    if (map_id < 0 or map_id >= maps_count or key_len > MAX_KEY_SIZE) {
        std.debug.print("[ZIG_MAPPER_ERROR]Invalid map id or key size {0},{1}\n", .{ map_id, key_len });
        return -1;
    }
    const map = maps[@intCast(map_id)] orelse return -1;
    const key_size: usize = @intCast(key_len);

    const hash_value = hash(key[0..key_size]);
    var current_entry = map[hash_value];
    while (current_entry) |curr_entry| {
        if (key_len == curr_entry.key.len and std.mem.eql(u8, curr_entry.key[0..key_size], key[0..key_size])) {
            std.mem.copy(u8, value[0..curr_entry.value.len], curr_entry.value[0..curr_entry.value.len]);
            return @intCast(curr_entry.value.len);
        }
        current_entry = curr_entry.next;
    }
    return -1;
}

export fn destroy_map(map_id: i32) void {
    if (map_id < 0 or map_id >= maps_count) {
        std.debug.print("[ZIG_MAPPER_ERROR]Invalid map id or key size: {}\n", .{map_id});
        return;
    }

    var map = maps[@intCast(map_id)] orelse return;

    std.debug.print("[ZIG_MAPPER_INFO]Destroying map {}\n", .{map_id});

    for (0..map.len) |i| {
        var current_entry = map[i];
        while (current_entry) |entry| {
            const next_entry = entry.next;
            allocator.free(entry.key);
            allocator.free(entry.value);
            allocator.destroy(entry);
            current_entry = next_entry;
        }
    }
    allocator.free(map);
    maps[@intCast(map_id)] = null;
}

export fn destroy_all_maps() void {
    for (0..maps_count) |i| {
        destroy_map(@intCast(i));
    }
    maps_count = 0;
}

test "test_init_map" {
    const map_id = init_map();
    try expect(map_id == 0);
    destroy_map(map_id);
}

test "test_set_val" {
    const map_id = init_map();
    const key = "key";
    const value = "value";
    const key_len: i32 = @intCast(key.len);
    const value_len: i32 = @intCast(value.len);
    const result = set_val(map_id, key, key_len, value, value_len);
    try expect(result != -1);
    destroy_map(map_id);
}

test "test_get_val" {
    const map_id = init_map();
    const key = "key";
    const value = "value";
    const key_len: i32 = @intCast(key.len);
    const value_len: i32 = @intCast(value.len);
    var r = set_val(map_id, &key[0], key_len, &value[0], value_len);
    try expect(r != -1);
    var result: [MAX_VALUE_SIZE]u8 = undefined;
    const get_result = get_val(map_id, &key[0], key_len, &result[0]);
    try expect(get_result != -1);
    const r_bytes: usize = @intCast(get_result);
    try expect(std.mem.eql(u8, result[0..r_bytes], value[0..r_bytes]) == true);
    destroy_map(map_id);
}

test "test_shadowing" {
    const map_id = init_map();
    const key = "key";
    const value = "value";
    const key_len: i32 = @intCast(key.len);
    const value_len: i32 = @intCast(value.len);
    var r = set_val(map_id, key, key_len, value, value_len);
    try expect(r != -1);
    const new_value = "new_value";
    const new_value_len: i32 = @intCast(new_value.len);
    r = set_val(map_id, key, key_len, new_value, new_value_len);
    try expect(r != -1);
    var result: [MAX_VALUE_SIZE]u8 = undefined;
    const get_result = get_val(map_id, key, key_len, &result[0]);
    try expect(get_result == new_value_len);
    try expect(std.mem.eql(u8, result[0..new_value_len], new_value[0..new_value_len]) == true);
    destroy_map(map_id);
}

File: src/mapper.c
#include "mapper.h"
#include "logger.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_KEY_SIZE 2048
#define MAX_VALUE_SIZE 10 << 20
#define HASH_MAP_SIZE 101
#define NUMBER_OF_MAPS 100
typedef char *value;
typedef const char *key;
typedef struct entry entry;

struct entry {
  key key;
  int key_len;
  value value;
  int value_len;
  entry *next;
};
typedef entry **hash_map_t;

hash_map_t maps[NUMBER_OF_MAPS];
int maps_count = 0;

int hash(key key, int key_len) {
  int hash = 0;
  for (int i = 0; i < key_len; i++) {
    hash += key[i];
  }
  hash = hash < 0 ? -hash : hash;
  return hash % HASH_MAP_SIZE;
}

int init_map() {
  if (maps_count >= NUMBER_OF_MAPS) {
    LOG_ERROR("Maximum number of maps reached");
    return -1;
  }
  maps[maps_count] = (hash_map_t)malloc(sizeof(entry *) * HASH_MAP_SIZE);
  for (int i = 0; i < HASH_MAP_SIZE; i++) {
    maps[maps_count][i] = NULL;
  }
  LOG_INFO("Creating new map %d", maps_count);
  return maps_count++;
}

int set_val(int map_id, const char *key, int key_len, const char *value,
            int value_len) {
  if (map_id < 0 || map_id >= maps_count || key_len > MAX_KEY_SIZE ||
      value_len > MAX_VALUE_SIZE) {
    LOG_ERROR("Invalid map_id or key/value size");
    return -1;
  }
  int hash_key = hash(key, key_len);
  entry *new_entry = (entry *)malloc(sizeof(entry));
  char *key_copy = (char *)malloc(key_len);
  char *value_copy = (char *)malloc(value_len);
  memcpy(key_copy, key, key_len);
  memcpy(value_copy, value, value_len);
  new_entry->key_len = key_len;
  new_entry->value_len = value_len;
  new_entry->key = key_copy;
  new_entry->value = value_copy;
  new_entry->next = NULL;

  LOG_DEBUG("Creating new entry on map %d", map_id);
  LOG_DEBUG("\thash_key: %d", hash_key);
  LOG_DEBUG("\tkey: -%*s-", key_len, key);
  LOG_DEBUG("\tvalue: -%*s-\n", value_len, value);
  entry *current = maps[map_id][hash_key];
  if (current == NULL) {
    maps[map_id][hash_key] = new_entry;
    return 0;
  } else if (key_len == current->key_len &&
             memcmp(current->key, key, key_len) == 0) {
    memcpy(current->value, value, value_len);
    current->value_len = value_len;
    free(key_copy);
    free(value_copy);
    free(new_entry);
    return 0;
  }
  while (current->next != NULL) {
    current = current->next;
  }
  current->next = new_entry;
  return hash_key;
}

int get_val(int map_id, const char *key, int key_len, char *value) {
  if (map_id < 0 || map_id >= maps_count) {
    LOG_ERROR("map_id out of range\n");
    LOG_ERROR("\tmap_id: %d\n", map_id);
    LOG_ERROR("\tmaps_count: %d\n", maps_count);
    return -1;
  }
  int hash_key = hash(key, key_len);

  LOG_DEBUG("Getting value on map %d\n", map_id);
  LOG_DEBUG("\thash_key: %d\n", hash_key);
  LOG_DEBUG("\tkey: -%*s-\n", key_len, key);
  entry *current = maps[map_id][hash_key];
  while (current != NULL) {
    if (key_len == current->key_len &&
        memcmp(current->key, key, key_len) == 0) {
      memcpy(value, current->value, current->value_len);
      return current->value_len;
    }
    current = current->next;
  }
  return -1;
}

void destroy_map(int map_id) {
  if (map_id < 0 || map_id >= maps_count) {
    LOG_ERROR("map_id out of range: %d\n", map_id);
    return;
  }
  LOG_INFO("Destroying map %d\n", map_id);
  for (int i = 0; i < HASH_MAP_SIZE; i++) {
    entry *current = maps[map_id][i];
    while (current != NULL) {
      entry *next = current->next;
      free((void *)current->key);
      free((void *)current->value);
      free(current);
      current = next;
    }
  }
  free(maps[map_id]);
  maps[map_id] = NULL;
}

void destroy_all_maps() {
  for (int i = 0; i < maps_count; i++) {
    if (maps[i] == NULL) {
      continue;
    }
    destroy_map(i);
  }
  maps_count = 0;
}

#ifdef TEST
int main() {
  int my_map = init_map();
  set_val(my_map, "key1", 4, "value1", 6);
  set_val(my_map, "key2", 4, "value2", 6);
  set_val(my_map, "key3", 4, "value3", 6);
  set_val(my_map, "key3", 4, "value9", 6);

  char buffer[100];
  int len = get_val(my_map, "key1", 4, buffer);
  printf("key1: %*s\n", len, buffer);
  len = get_val(my_map, "key2", 4, buffer);
  printf("key2: %*s\n", len, buffer);
  len = get_val(my_map, "key3", 4, buffer);
  printf("key3: %*s\n", len, buffer);
  destroy_map(my_map);

  return 0;
}
#endif
